File: app/src/test/java/com/beautifulquran/ui/theme/ornament/OrnamentGeneratorTest.kt
package com.beautifulquran.ui.theme.ornament

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

class OrnamentGeneratorTest {

    /**
     * mulberry32 known-answer values (computed with the reference JS
     * implementation). The web port asserts the same values — this is the
     * cross-platform contract that keeps both covers drawing from one stream.
     */
    @Test
    fun `prng matches reference mulberry32 stream`() {
        val expected = mapOf(
            1 to longArrayOf(2693262067L, 11749833L, 2265367787L, 4213581821L),
            123456789 to longArrayOf(1107202814L, 4169434471L, 3372958138L, 885470128L),
            -7 to longArrayOf(1860010037L, 1397564179L, 2337619704L, 2062400319L),
        )
        for ((seed, values) in expected) {
            val rng = Mulberry32(seed)
            for (v in values) assertEquals("seed $seed", v, rng.nextUInt())
        }
    }

    @Test
    fun `same seed grows the same ornament`() {
        for (seed in intArrayOf(1, 42, -913, 2_000_000_011.toInt())) {
            assertEquals(generateCoverOrnament(seed), generateCoverOrnament(seed))
        }
    }

    @Test
    fun `different seeds grow different ornaments`() {
        val a = generateCoverOrnament(1)
        var anyDiffer = false
        for (seed in 2..12) {
            if (generateCoverOrnament(seed) != a) anyDiffer = true
        }
        assertTrue(anyDiffer)
    }

    @Test
    fun `every seed in a wide sample generates a sane ornament`() {
        for (seed in 0 until 300) {
            val o = generateCoverOrnament(seed * 7919 + seed)
            assertTrue(o.medallion.strokes.size >= 4)
            assertTrue(o.medallion.dots.isNotEmpty())
            assertTrue(o.cornerSeal.strokes.isNotEmpty())
            assertTrue(o.border.strokes.isNotEmpty())
            assertTrue(o.border.period > 0.5)
            assertTrue(o.field.strokes.size >= 8)
            assertTrue(o.field.cellW > 0 && o.field.cellH > 0)
            assertTrue(o.field.cellWidthDp in 40.0..200.0)

            for (s in o.medallion.strokes + o.cornerSeal.strokes) {
                assertTrue(s.points.size >= 2)
                assertTrue(s.birth >= 0.0 && s.birth + s.span <= 1.0001)
                for (p in s.points) {
                    assertTrue("medallion point out of unit box: $p", p.x in -0.001..1.001)
                    assertTrue("medallion point out of unit box: $p", p.y in -0.001..1.001)
                }
            }
            for (s in o.border.strokes) {
                for (p in s.points) {
                    assertTrue("border x outside period: $p", p.x in -0.001..o.border.period + 0.001)
                    assertTrue("border y outside band: $p", p.y in -0.001..1.001)
                }
            }
            // Field strokes stay near their cell (Hankin keeps rays inside
            // each polygon; polygons may legitimately straddle cell edges).
            val fx = 1.5 * o.field.cellW
            val fy = 1.5 * o.field.cellH
            for (s in o.field.strokes) {
                for (p in s.points) {
                    assertTrue("field point far outside cell: $p", p.x in -fx..fx + o.field.cellW)
                    assertTrue("field point far outside cell: $p", p.y in -fy..fy + o.field.cellH)
                }
            }
        }
    }

    @Test
    fun `medallion has full n-fold rotational symmetry`() {
        for (seed in intArrayOf(3, 17, 99, 1234, -55, 777_777)) {
            val m = generateCoverOrnament(seed).medallion
            val angle = 2.0 * PI / m.fold
            val pts = m.strokes.flatMap { it.points }
            for (p in pts) {
                val rx = 0.5 + (p.x - 0.5) * cos(angle) - (p.y - 0.5) * sin(angle)
                val ry = 0.5 + (p.x - 0.5) * sin(angle) + (p.y - 0.5) * cos(angle)
                val hit = pts.any { q -> abs(q.x - rx) < 1e-6 && abs(q.y - ry) < 1e-6 }
                assertTrue("seed $seed fold ${m.fold}: rotated point unmatched ($rx, $ry)", hit)
            }
        }
    }

    @Test
    fun `field pattern is continuous across cell edges`() {
        // Every Hankin stroke starts at an edge midpoint shared with the
        // neighbouring polygon (possibly in the next cell), so each start
        // point must be matched by another stroke start at the same spot,
        // modulo the cell period.
        for (seed in intArrayOf(5, 8, 21, 100, 4242)) {
            val f = generateCoverOrnament(seed).field
            val starts = f.strokes.map { it.points.first() }
            for (s in starts) {
                var mates = 0
                for (t in starts) {
                    for (dx in -1..1) {
                        for (dy in -1..1) {
                            val tx = t.x + dx * f.cellW
                            val ty = t.y + dy * f.cellH
                            if (abs(tx - s.x) < 1e-6 && abs(ty - s.y) < 1e-6) mates++
                        }
                    }
                }
                // The ray itself plus at least its sibling ray; midpoints on
                // shared edges collect all four rays of the two polygons.
                assertTrue("seed $seed: lonely ray at (${s.x}, ${s.y})", mates >= 2)
            }
        }
    }

    @Test
    fun `never draws a hexagram - no triangles, no 6-fold stars, anywhere`() {
        for (seed in 0 until 400) {
            val o = generateCoverOrnament(seed * 104729 + 13)
            assertTrue("seal fold 6 (seed $seed)", o.cornerSeal.fold != 6)
            val everyStroke = o.medallion.strokes + o.cornerSeal.strokes +
                o.border.strokes + o.field.strokes
            for (s in everyStroke) {
                assertTrue(
                    "closed triangle found (seed $seed)",
                    !(s.closed && s.points.size == 3),
                )
            }
        }
    }

    @Test
    fun `variety - a seed sample uses every fold, motif, tiling and border`() {
        val folds = HashSet<Int>()
        val tilingCells = HashSet<Long>()
        val borderShapes = HashSet<Int>()
        for (seed in 0 until 200) {
            val o = generateCoverOrnament(seed)
            folds.add(o.medallion.fold)
            tilingCells.add((o.field.cellW * 1000).toLong() * 100_000 + (o.field.cellH * 1000).toLong())
            borderShapes.add(o.border.strokes.size * 31 + o.border.dots.size)
        }
        assertEquals(setOf(8, 10, 12, 16), folds)
        assertEquals(3, tilingCells.size)
        assertTrue("expected several border grammars, got $borderShapes", borderShapes.size >= 3)
    }
}

File: web/src/ui/theme/__tests__/ornamentGenerator.test.ts
import { describe, expect, it } from 'vitest'
import {
  chapterOrnamentSeed,
  generateChapterOrnament,
  generateCoverOrnament,
  Mulberry32,
  SEAL_RING_RADIUS,
  type OrnamentPoint,
} from '../ornamentGenerator'

describe('ornamentGenerator', () => {
  it('prng matches the reference mulberry32 stream (Android parity)', () => {
    // Same known-answer values asserted by OrnamentGeneratorTest.kt — the
    // cross-platform contract that keeps both covers drawing from one stream.
    const expected: Array<[number, number[]]> = [
      [1, [2693262067, 11749833, 2265367787, 4213581821]],
      [123456789, [1107202814, 4169434471, 3372958138, 885470128]],
      [-7, [1860010037, 1397564179, 2337619704, 2062400319]],
    ]
    for (const [seed, values] of expected) {
      const rng = new Mulberry32(seed)
      for (const v of values) expect(rng.nextUInt()).toBe(v)
    }
  })

  it('grows the same ornament from the same seed', () => {
    for (const seed of [1, 42, -913, 2000000011]) {
      expect(generateCoverOrnament(seed)).toEqual(generateCoverOrnament(seed))
    }
  })

  it('grows different ornaments from different seeds', () => {
    const a = generateCoverOrnament(1)
    let anyDiffer = false
    for (let seed = 2; seed <= 12; seed++) {
      try {
        expect(generateCoverOrnament(seed)).not.toEqual(a)
        anyDiffer = true
      } catch {
        /* one collision is tolerable; all colliding is not */
      }
    }
    expect(anyDiffer).toBe(true)
  })

  it('generates a sane ornament for a wide seed sample', () => {
    // Plain checks collected into a violation list (per-point expect() calls
    // are far too slow at this volume); a single assert reports the first few.
    const violations: string[] = []
    const check = (ok: boolean, label: string) => {
      if (!ok && violations.length < 5) violations.push(label)
    }
    for (let seed = 0; seed < 300; seed++) {
      const o = generateCoverOrnament(seed * 7919 + seed)
      const at = `seed ${seed * 7919 + seed}`
      check(o.medallion.strokes.length >= 4, `${at}: medallion strokes`)
      check(o.medallion.dots.length > 0, `${at}: medallion dots`)
      check(o.cornerSeal.strokes.length > 0, `${at}: seal strokes`)
      check(o.border.strokes.length > 0, `${at}: border strokes`)
      check(o.border.period > 0.5, `${at}: border period`)
      check(o.field.strokes.length >= 8, `${at}: field strokes`)
      check(o.field.cellW > 0 && o.field.cellH > 0, `${at}: field cell`)

      // The medallion has no bezel; the seal's petal tips must aim down
      // the band axes just past the ring, and its bezel may reach past
      // the unit box by at most (tip − 0.5).
      check(o.medallion.tipRadius === 0, `${at}: medallion tipRadius`)
      check(
        o.cornerSeal.tipRadius > SEAL_RING_RADIUS && o.cornerSeal.tipRadius <= 0.7,
        `${at}: seal tipRadius ${o.cornerSeal.tipRadius}`,
      )
      for (const s of o.medallion.strokes) {
        check(s.points.length >= 2, `${at}: short stroke`)
        check(s.birth >= 0 && s.birth + s.span <= 1.0001, `${at}: build window`)
        for (const p of s.points) {
          check(
            p.x >= -0.001 && p.x <= 1.001 && p.y >= -0.001 && p.y <= 1.001,
            `${at}: medallion point outside unit box (${p.x}, ${p.y})`,
          )
        }
      }
      for (const s of o.cornerSeal.strokes) {
        check(s.points.length >= 2, `${at}: short stroke`)
        check(s.birth >= 0 && s.birth + s.span <= 1.0001, `${at}: build window`)
        for (const p of s.points) {
          check(
            p.x >= -0.2 && p.x <= 1.2 && p.y >= -0.2 && p.y <= 1.2,
            `${at}: seal point too far out (${p.x}, ${p.y})`,
          )
        }
      }
      // The khatam chain's link diamond straddles the period boundary by
      // design (its far half is completed by the neighbouring tile when
      // the band repeats), so the x margin allows for it; y always stays
      // inside the band.
      for (const s of o.border.strokes) {
        for (const p of s.points) {
          check(
            p.x >= -0.2 && p.x <= o.border.period + 0.2 && p.y >= -0.001 && p.y <= 1.001,
            `${at}: border point outside band (${p.x}, ${p.y})`,
          )
        }
      }
    }
    expect(violations).toEqual([])
  })

  it('medallion has full n-fold rotational symmetry', () => {
    for (const seed of [3, 17, 99, 1234, -55, 777777]) {
      const m = generateCoverOrnament(seed).medallion
      const angle = (2 * Math.PI) / m.fold
      const pts: OrnamentPoint[] = m.strokes.flatMap((s) => s.points)
      for (const p of pts) {
        const rx = 0.5 + (p.x - 0.5) * Math.cos(angle) - (p.y - 0.5) * Math.sin(angle)
        const ry = 0.5 + (p.x - 0.5) * Math.sin(angle) + (p.y - 0.5) * Math.cos(angle)
        const hit = pts.some((q) => Math.abs(q.x - rx) < 1e-6 && Math.abs(q.y - ry) < 1e-6)
        expect(hit, `seed ${seed} fold ${m.fold}: rotated point unmatched`).toBe(true)
      }
    }
  })

  it('field pattern is continuous across cell edges', () => {
    for (const seed of [5, 8, 21, 100, 4242]) {
      const f = generateCoverOrnament(seed).field
      const starts = f.strokes.map((s) => s.points[0]!)
      for (const s of starts) {
        let mates = 0
        for (const t of starts) {
          for (let dx = -1; dx <= 1; dx++) {
            for (let dy = -1; dy <= 1; dy++) {
              const tx = t.x + dx * f.cellW
              const ty = t.y + dy * f.cellH
              if (Math.abs(tx - s.x) < 1e-6 && Math.abs(ty - s.y) < 1e-6) mates++
            }
          }
        }
        expect(mates, `seed ${seed}: lonely ray`).toBeGreaterThanOrEqual(2)
      }
    }
  })

  it('never draws a hexagram — no triangles, no 6-fold stars, anywhere', () => {
    const violations: string[] = []
    for (let seed = 0; seed < 400; seed++) {
      const o = generateCoverOrnament(seed * 104729 + 13)
      if (o.cornerSeal.fold === 6) violations.push(`seed ${seed}: 6-fold seal`)
      const everyStroke = [
        ...o.medallion.strokes,
        ...o.cornerSeal.strokes,
        ...o.border.strokes,
        ...o.field.strokes,
      ]
      for (const s of everyStroke) {
        if (s.closed && s.points.length === 3) violations.push(`seed ${seed}: triangle`)
      }
    }
    expect(violations).toEqual([])
  })

  it('a seed sample uses every fold, motif family, tiling, and border', () => {
    const folds = new Set<number>()
    const tilings = new Set<string>()
    const borders = new Set<string>()
    for (let seed = 0; seed < 200; seed++) {
      const o = generateCoverOrnament(seed)
      folds.add(o.medallion.fold)
      tilings.add(`${o.field.cellW.toFixed(3)}x${o.field.cellH.toFixed(3)}`)
      borders.add(`${o.border.strokes.length}/${o.border.dots.length}`)
    }
    expect([...folds].sort((a, b) => a - b)).toEqual([8, 10, 12, 16])
    expect(tilings.size).toBe(3)
    expect(borders.size).toBeGreaterThanOrEqual(3)
  })

  it('chapter seed recovers the chapter number regardless of ayah count', () => {
    // Chapters are numbered 1..114 (not 0-indexed), so the recovered digit
    // is ((seed - 1) mod 114) + 1, not a plain mod 114.
    for (let chapter = 1; chapter <= 114; chapter++) {
      for (const ayahCount of [3, 6, 11, 88, 286]) {
        const seed = chapterOrnamentSeed(chapter, ayahCount)
        expect(((seed - 1) % 114) + 1).toBe(chapter)
      }
    }
  })

  it('chapter seed is unique across all 114 chapters even at a shared ayah count', () => {
    // A real duplicate: 62, 63, 93, 100, 101 all have exactly 11 ayahs.
    const seeds = new Set<number>()
    for (let chapter = 1; chapter <= 114; chapter++) seeds.add(chapterOrnamentSeed(chapter, 11))
    expect(seeds.size).toBe(114)
  })

  it('reproduces the same ornament for the same chapter and ayah count', () => {
    const seed = chapterOrnamentSeed(2, 286)
    expect(generateChapterOrnament(seed)).toEqual(generateChapterOrnament(seed))
  })

  it('renders different rosettes and fields for chapters that share an ayah count', () => {
    const elevenAyahChapters = [62, 63, 93, 100, 101]
    const ornaments = elevenAyahChapters.map((c) => generateChapterOrnament(chapterOrnamentSeed(c, 11)))
    const uniqueRosettes = new Set(ornaments.map((o) => JSON.stringify(o.rosette)))
    const uniqueFields = new Set(ornaments.map((o) => JSON.stringify(o.field)))
    expect(uniqueRosettes.size).toBe(elevenAyahChapters.length)
    expect(uniqueFields.size).toBe(elevenAyahChapters.length)
  })

  it('chapter ornament never draws a hexagram', () => {
    const violations: string[] = []
    for (let chapter = 1; chapter <= 114; chapter++) {
      for (const ayahCount of [3, 6, 11, 88, 286]) {
        const ornament = generateChapterOrnament(chapterOrnamentSeed(chapter, ayahCount))
        for (const s of [...ornament.rosette.strokes, ...ornament.field.strokes]) {
          if (s.closed && s.points.length === 3) {
            violations.push(`chapter ${chapter}, ${ayahCount} ayahs: triangle`)
          }
        }
      }
    }
    expect(violations).toEqual([])
  })
})
